1387 criteria reader doesn't respect getter prefixes

SpyrosKou opened the pull request:
This is an attempt to fix autogenerated Criteria that do not work in `find()` in the `MongoBackend` an issue also described in https://github.com/immutables/immutables/issues/1387 with more details.
The conversion from the property names to the MongoDB fields seems to take place in `MongoSession.java`.
There the `_id` property was handled individually and all other properties were delegated to `PathNaming.defaultNaming()` which returned the path as a field name.  This did not work in MongoDB filter queries. As for example it returned `getDescription` instead of `description` and the filter query required the field name, being `description`.
Replacing the `PathNaming.defaultNaming()` with the `JavaBeanNaming()` provides accurate field name transitions.

2 tests have also been included that test that `find() `works with the `ID`, with a getter of the form `getProperty()' and with a getter of the form `property()'.  I named properties more specifically to facilitate any potential future references.

PS: This PR solves some cases and improves the situation, but more changes should be examined. For instance taking into account the `@JsonProperty` annotation.

SpyrosKou committed:
#1387 add some tests that demonstrate the problem

This is a bit draft location of the tests, most probably these tests should be relocated or deleted.

SpyrosKou committed:
#1387 Use JavaBeanNaming in criteria framework instead of PathNaming.defaultNaming()

Using PathNaming.defaultNaming() in MongoPathNaming will result in all property names being the member path.

elucash commented:
@asereda-gs can you check this when you have time. I would just merge it, but need to understand how well it addresses the underlying problem and if it causes any issues / incompatibilities

SpyrosKou committed:
#1387 use mongo-java-server instead of mongo, reduced test LoC

-Use mongo-java-server to allow github actions to run and to follow projects suggestions.
-Removed duplication in code.
-Added javadoc

asereda-gs commented:
@elucash let me check this PR on weekend. Sorry was busy lately.

asereda-gs reviewed (commented):
Hello.

Thank you for this PR!

I'm thinking if allowing users to register their own `PathNaming` instance in `MongoSetup` would be a more flexible strategy?


Overall `PathNaming` needs to be better integrated with Gson/Jackson/Bson serializations. 

I assume this issue also affects ElasticSearch backend?

asereda-gs commented:
Ideally one should re-use existing  BSON `Codec` which is being employed when entities are being persisted. The problem is that we can't know the path mapping unless the entity is serialized.

asereda-gs commented:
I have prepared some ideas in a separate branch: https://github.com/immutables/immutables/tree/jackson-pathnaming

it needs some more work

asereda-gs commented:
Please take a look at 3fb8e3 and let me know if it works for you.
